### DISCUSSION TITLE
Token indices sequence length is longer than the specified maximum sequence length for this model (1538 > 512)

### DISCUSSION THREAD

**Discussion by wirokemproh** [aub-mind/arabert#72] [Q&A] [answered]
I'm trying to do inference. Here is the code
```
model = 'aubmindlab/bert-base-arabertv2'

arabert_prep = ArabertPreprocessor(model_name=model)
config = AutoConfig.from_pretrained(model)
tokenizer = AutoTokenizer.from_pretrained(model, do_lower_case=False)
model = AutoModelForSequenceClassification.from_pretrained(model, config=config)
model.eval()

df = pd.read_csv("../dataset/processed/sentiment_ar_neg.csv")
sentences = df['text'].to_list()

for sentence in sentences:
    text_preprocessed = arabert_prep.preprocess(sentence)
    print(text_preprocessed)

    input = tokenizer.encode(text_preprocessed,add_special_tokens=True, return_tensors='pt')

    #input = parse_and_tokenize(text_preprocessed)

    with torch.no_grad():
        pred = model(input)
        print(pred)
        pred = pred[0]
    outputs = pred.numpy()
    scores = np.exp(outputs) / np.exp(outputs).sum(-1, keepdims=True)
    result = [
        {"label": label[item.argmax()], "score": item.max().item()} for item in scores
    ]
    print(result)
```

Is this the correct way to do inference with Arabert?

I got an error like this on one of the data:
```
Token indices sequence length is longer than the specified maximum sequence length for this model (1538 > 512). Running this sequence through the model will result in indexing errors
Traceback (most recent call last):
```

Do I need to truncate the text, first?
Or do I miss something else?

**Comment by WissamAntoun** [comment, marked as answer]
I recommend you use the `pipeline` class from hugging face like this:
```python
from transformers import AutoModelForSequenceClassification, AutoTokenizer, pipeline

folder_name= "PATH TO THE FOLDER WHERE YOU SAVE THE FINETUNED MODEL"
tokenizer = AutoTokenizer.from_pretrained(folder_name)
model = AutoModelForSequenceClassification.from_pretrained(folder_name)

classifier = pipeline(
    "sentiment-analysis",
    model = model,
    tokenizer = tokenizer,
    device=0 # 0 for CUDA, -1 for CPU
)

preds = classifier(List_of_text)
```

You need to make sure that the tokenizer is saved with the model by using `tokenizer.save_pretrained("PATH")`, or you can re-use the tokenizer of the original model, but the first option makes it easier. You can also edit the config file to let the pipeline directly output class names by editing `id2label` and `label2id` dicts (preferably before you save the model) like this if you were using the `trainer` class from `HuggingFace`

```python
label_map = {'Positive': 0, 'Neutral': 1, 'Negative': 2}
inv_label_map = { v:k for k, v in label_map.items()}
trainer.model.config.label2id = label_map
trainer.model.config.id2label = inv_label_map
trainer.save_model("PATH")
```


Now as for the length you will need to truncate the input sentence to something lower than 512, or you can split the sentence into multiple examples and then recombine them in someway that takes into account the score and label of each splitted sentence

**Reply by wirokemproh** [reply]
Thanks @WissamAntoun .

I have further question.  Could you advice how do I truncate the sentence? 
Is 512 the length of the sentence? Or is it the number of word in the sentence?

when I saw this error:
```
RuntimeError: The size of tensor a (517) must match the size of tensor b (512) at non-singleton dimension 1
```

the length of the sentence is actually 1859, instead of 517

```
len(sentences[i])
Out[3]: 1589
input.size()
Out[4]: torch.Size([1, 517])
```

**Reply by WissamAntoun** [reply]
I think the tokenizer can split the sentence and return multiple examples from 1 input sentence, try to check that out

**Reply by wirokemproh** [reply]
Thanks @WissamAntoun . Works good now
